spark-cassandra-connector/src/main/scala/org/apache/spark/sql/cassandra/CassandraSourceRelation.scala

```diff
@@ -0,0 +1,201 @@
+
+  protected[this] val columnNameByLowercase = tableDef.allColumns
+    .map(c => (c.columnName.toLowerCase, c.columnName)).toMap
+
+  override def schema: StructType = {
+    def columnToStructField(column: ColumnDef): StructField = {
+      StructField(column.columnName, ColumnDataType.catalystDataType(column.columnType, nullable = true))
+    }
+    if (userSpecifiedSchema.isDefined){
+      userSpecifiedSchema.get
+    } else {
+      StructType(tableDef.allColumns.map(columnToStructField))
+    }
+  }
+
+  protected[this] val baseRdd = sqlContext
+    .sparkContext
+    .cassandraTable[CassandraSQLRow](keyspace, table)(
+      new CassandraConnector(sqlContext.getCassandraConnConf(cluster)),
+      sqlContext.getReadConf(keyspace, table, cluster),
+      implicitly[ClassTag[CassandraSQLRow]],
+      CassandraSQLRowReader,
+      implicitly[ValidRDDType[CassandraSQLRow]])
+
+  private[this] val writeConf = sqlContext.getWriteConf(keyspace, table, cluster)
+
+  override def insert(data: DataFrame, overwrite: Boolean): Unit = {
+    data.rdd.saveToCassandra(keyspace, table, AllColumns, writeConf)(
+        new CassandraConnector(sqlContext.getCassandraConnConf(cluster)), SqlRowWriter.Factory)
+  }
+
+  def buildScan(): RDD[Row] = baseRdd.asInstanceOf[RDD[Row]]
+
+  protected[this] def getSchemaData(column: String, row: Row): (Any, NativeType) = {
+    val index = row.asInstanceOf[CassandraSQLRow].indexOf(column)
+    val columnDef = tableDef.columnByName.get(column).getOrElse(throw new IOException(s"Can't find $column"))
+    val dataType = ColumnDataType.catalystDataType(columnDef.columnType, nullable = true)
+    require(dataType.isPrimitive, s"${dataType.typeName} is not supported in filter.")
+
+    (row.get(index), dataType.asInstanceOf[NativeType])
+  }
+}
+
+/** Table scan relation implements [[BaseRelation]], [[InsertableRelation]] and [[TableScan]] */
+private[cassandra] class TableScanRelationImpl(table: String, keyspace: String, cluster: Option[String],
+                                               userSpecifiedSchema: Option[StructType], sqlcontext: SQLContext)
+  extends BaseRelationImpl(table, keyspace, cluster, userSpecifiedSchema, sqlcontext) with TableScan {
+  override def buildScan(): RDD[Row] = super.buildScan()
+}
+
+/** Table scan relation implements [[BaseRelation]], [[InsertableRelation]] and [[PrunedScan]] */
+private[cassandra] class PrunedScanRelationImpl(table: String, keyspace: String, cluster: Option[String],
+                                                userSpecifiedSchema: Option[StructType], sqlcontext: SQLContext)
+  extends BaseRelationImpl(table, keyspace, cluster, userSpecifiedSchema, sqlcontext) with PrunedScan {
+  override def buildScan(requiredColumns: Array[String]): RDD[Row] = {
+    val transformer = new RddFilterSelectPdTrf(requiredColumns, columnNameByLowercase, Seq.empty)
+    (transformer.maybeSelect apply baseRdd).asInstanceOf[RDD[Row]]
```

Comment:
hmm..
`(transformer.maybeSelect apply baseRdd).asInstanceOf[RDD[Row]]`
vs
`transformer.maybeSelect(baseRdd)`
Which one is more readable?